src/java/org/apache/cassandra/journal/Flusher.java

```diff
@@ -97,20 +102,144 @@ void start()
+            public void run(Interruptible.State state) throws InterruptedException
+            {
+                try
+                {
+                    doRun(state);
+                }
+                catch (Throwable t)
+                {
+                    if (!journal.handleError("Failed to flush segments to disk", t))
+                        throw new TerminateException();
+                }
+            }
+
+            private void awaitWork() throws InterruptedException
+            {
+                long lastStartedAt = fsyncStartedFor;
+                if (fsyncWaitingSince != lastStartedAt)
+                    return;
+
+                awaitingWork = Thread.currentThread();
+                do
+                {
+                    if (Thread.interrupted())
+                    {
+                        awaitingWork = null;
+                        throw new InterruptedException();
+                    }
+
+                    LockSupport.park();
+                }
+                while (fsyncWaitingSince == lastStartedAt);
+
+                awaitingWork = null;
+            }
+
+            void notify(Thread notify)
+            {
+                if (notify != null)
+                    LockSupport.unpark(notify);
+            }
+
+            public void doRun(Interruptible.State state) throws InterruptedException
+            {
+                awaitWork();
+                if (fsyncing == null)
+                    fsyncing = journal.oldestActiveSegment();
+
+                // invert order of access; we might see a future fsyncTo, but at worst this means redundantly invoking fsync before updating fsyncStartedFor
+                long startedAt = fsyncWaitingSince;
+                ActiveSegment<K, V> fsyncTo = this.fsyncTo;
+                fsyncStartedFor = startedAt;
+                // synchronized to prevent thread interrupts while performing IO operations and also
+                // clear interrupted status to prevent ClosedByInterruptException in ActiveSegment::flush
+                synchronized (this)
+                {
+                    boolean ignore = Thread.interrupted();
+                    while (fsyncing != fsyncTo)
+                    {
+                        fsyncing.fsync();
+                        journal.closeActiveSegmentAndOpenAsStatic(fsyncing);
+                        fsyncing = journal.getActiveSegment(fsyncing.descriptor.timestamp + 1);
+                    }
+                    fsyncing.fsync();
+                }
+                fsyncFinishedFor = startedAt;
+                flushComplete.signalAll();
+                long finishedAt = clock.now();
+                processDuration(startedAt, finishedAt);
+            }
+
+            void postFlush(long startedAt, ActiveSegment<K, V> segment, int syncedOffset) throws InterruptedException
+            {
+                long requireFsyncTo = startedAt - periodicFlushLagBlockNanos();
+
+                fsyncTo = segment;
```

Comment:
nit: I understand that this is an additional allocation; but maybe we could make these two fields a single variable? 

An alternative could be a sentinel boolean that would make another thread spin.